File: sdk/typescript/src/builder/tree.ts
import type { ViewNode, ViewTree } from "../generated/wit.ts";
import type { ViewNodeSpec } from "./nodes.ts";

export type { ViewNode, ViewTree };

/**
 * Builds a flattened, valid `ViewTree` from a root `ViewNodeSpec` tree structure.
 *
 * Flattens all nested child nodes into a linear `nodes: ViewNode[]` list with 0-indexed
 * root and child reference arrays. Preserves all zero, false, and empty values accurately.
 */
export function buildViewTree(rootSpec: ViewNodeSpec): ViewTree {
  const nodes: ViewNode[] = [];

  function addNode(spec: ViewNodeSpec): number {
    const nodeIndex = nodes.length;

    // Reserve spot for this node
    switch (spec.tag) {
      case "container": {
        const childIndices: number[] = [];
        // Temporary placeholder
        nodes.push({
          tag: "container",
          val: {
            direction: spec.val.direction ?? { tag: "column" },
            children: new Uint32Array(0),
            style: spec.val.style,
            gap: spec.val.gap,
            alignItems: spec.val.alignItems,
            justifyContent: spec.val.justifyContent,
            wrap: spec.val.wrap ?? false,
            eventId: spec.val.eventId,
          },
        });

        if (spec.val.children) {
          for (const childSpec of spec.val.children) {
            childIndices.push(addNode(childSpec));
          }
        }

        // Update with resolved child indices
        nodes[nodeIndex] = {
          tag: "container",
          val: {
            direction: spec.val.direction ?? { tag: "column" },
            children: new Uint32Array(childIndices),
            style: spec.val.style,
            gap: spec.val.gap,
            alignItems: spec.val.alignItems,
            justifyContent: spec.val.justifyContent,
            wrap: spec.val.wrap ?? false,
            eventId: spec.val.eventId,
          },
        };
        break;
      }

      case "list": {
        const itemIndices: number[] = [];
        nodes.push({
          tag: "list",
          val: {
            items: new Uint32Array(0),
            style: spec.val.style,
          },
        });

        if (spec.val.items) {
          for (const itemSpec of spec.val.items) {
            itemIndices.push(addNode(itemSpec));
          }
        }

        nodes[nodeIndex] = {
          tag: "list",
          val: {
            items: new Uint32Array(itemIndices),
            style: spec.val.style,
          },
        };
        break;
      }

      case "text":
        nodes.push({ tag: "text", val: { ...spec.val } });
        break;

      case "icon":
        nodes.push({ tag: "icon", val: { ...spec.val } });
        break;

      case "image":
        nodes.push({ tag: "image", val: { ...spec.val } });
        break;

      case "button":
        nodes.push({ tag: "button", val: { ...spec.val } });
        break;

      case "icon-button":
        nodes.push({ tag: "icon-button", val: { ...spec.val } });
        break;

      case "toggle":
        nodes.push({ tag: "toggle", val: { ...spec.val } });
        break;

      case "slider":
        nodes.push({ tag: "slider", val: { ...spec.val } });
        break;

      case "text-input":
        nodes.push({ tag: "text-input", val: { ...spec.val } });
        break;

      case "spacer":
        nodes.push({ tag: "spacer", val: { ...spec.val } });
        break;

      case "divider":
        nodes.push({ tag: "divider" });
        break;

      case "badge":
        nodes.push({ tag: "badge", val: { ...spec.val } });
        break;

      case "progress":
        nodes.push({ tag: "progress", val: { ...spec.val } });
        break;

      case "loading-indicator":
        nodes.push({ tag: "loading-indicator", val: { ...spec.val } });
        break;
    }

    return nodeIndex;
  }

  const root = addNode(rootSpec);
  return { nodes, root };
}

File: sdk/rust/src/builder/tree.rs
//! Tree flattening and canonical ViewTree assembly.

use crate::bindings::shilpo::extension::view::{
    BadgeNode, ButtonNode, ContainerNode, IconButtonNode, IconNode, ImageNode, ListNode,
    LoadingIndicatorNode, ProgressNode, SliderNode, SpacerNode, TextInputNode, TextNode,
    ToggleNode, ViewNode, ViewTree,
};
use crate::builder::nodes::{IntoViewNode, NodeSpec};

/// Flattens a root node specification tree into a canonical indexed [`ViewTree`].
///
/// Preserves child order deterministically, sets the root index to 0, and
/// populates child index arrays for container and list nodes.
///
/// # Examples
///
/// ```rust
/// use shilpo_ext_sdk::prelude::*;
///
/// let tree = build_view_tree(
///     row()
///         .gap(8.0)
///         .child(icon("star").size(16.0))
///         .child(text("Favorite"))
/// );
///
/// assert_eq!(tree.root, 0);
/// assert_eq!(tree.nodes.len(), 3);
/// ```
pub fn build_view_tree(root: impl IntoViewNode) -> ViewTree {
    let mut nodes = Vec::new();
    let root_spec = root.into_node_spec();
    flatten_node_spec(root_spec, &mut nodes);
    ViewTree { nodes, root: 0 }
}

fn flatten_node_spec(spec: NodeSpec, nodes: &mut Vec<ViewNode>) -> u32 {
    let current_index = nodes.len() as u32;

    match spec {
        NodeSpec::Container(c) => {
            nodes.push(ViewNode::Divider);
            let mut child_indices = Vec::with_capacity(c.children.len());
            for child in c.children {
                child_indices.push(flatten_node_spec(child, nodes));
            }
            nodes[current_index as usize] = ViewNode::Container(ContainerNode {
                direction: c.direction,
                children: child_indices,
                style: c.style,
                gap: c.gap,
                align_items: c.align_items,
                justify_content: c.justify_content,
                wrap: c.wrap,
                event_id: c.event_id,
            });
        }
        NodeSpec::List(l) => {
            nodes.push(ViewNode::Divider);
            let mut item_indices = Vec::with_capacity(l.items.len());
            for item in l.items {
                item_indices.push(flatten_node_spec(item, nodes));
            }
            nodes[current_index as usize] = ViewNode::List(ListNode {
                items: item_indices,
                style: l.style,
            });
        }
        NodeSpec::Text(t) => {
            nodes.push(ViewNode::Text(TextNode {
                content: t.content,
                font_size: t.font_size,
                bold: t.bold,
                style: t.style,
            }));
        }
        NodeSpec::Icon(i) => {
            nodes.push(ViewNode::Icon(IconNode {
                name: i.name,
                size: i.size,
                style: i.style,
            }));
        }
        NodeSpec::Image(im) => {
            nodes.push(ViewNode::Image(ImageNode {
                asset_path: im.asset_path,
                width: im.width,
                height: im.height,
                style: im.style,
            }));
        }
        NodeSpec::Button(b) => {
            nodes.push(ViewNode::Button(ButtonNode {
                label: b.label,
                event_id: b.event_id,
                style: b.style,
            }));
        }
        NodeSpec::IconButton(ib) => {
            nodes.push(ViewNode::IconButton(IconButtonNode {
                icon_name: ib.icon_name,
                event_id: ib.event_id,
                style: ib.style,
            }));
        }
        NodeSpec::Toggle(tog) => {
            nodes.push(ViewNode::Toggle(ToggleNode {
                value: tog.value,
                event_id: tog.event_id,
                style: tog.style,
            }));
        }
        NodeSpec::Slider(s) => {
            nodes.push(ViewNode::Slider(SliderNode {
                value: s.value,
                min: s.min,
                max: s.max,
                event_id: s.event_id,
                style: s.style,
            }));
        }
        NodeSpec::TextInput(ti) => {
            nodes.push(ViewNode::TextInput(TextInputNode {
                placeholder: ti.placeholder,
                value: ti.value,
                event_id: ti.event_id,
                style: ti.style,
            }));
        }
        NodeSpec::Spacer(sp) => {
            nodes.push(ViewNode::Spacer(SpacerNode { size: sp.size }));
        }
        NodeSpec::Divider => {
            nodes.push(ViewNode::Divider);
        }
        NodeSpec::Badge(bg) => {
            nodes.push(ViewNode::Badge(BadgeNode {
                label: bg.label,
                style: bg.style,
            }));
        }
        NodeSpec::Progress(p) => {
            nodes.push(ViewNode::Progress(ProgressNode {
                value: p.value,
                style: p.style,
            }));
        }
        NodeSpec::LoadingIndicator(li) => {
            nodes.push(ViewNode::LoadingIndicator(LoadingIndicatorNode {
                size: li.size,
                color: li.color,
                style: li.style,
            }));
        }
    }

    current_index
}
